File: java/GenericArray.java
import java.util.*;

public class GenericArray<T> {
    // Use Object[] to store generic elements
    private T[] data;

    // Constructor to initialize the array
    @SuppressWarnings("unchecked")
    public GenericArray(T[] initialData) {
        this.data = (T[]) new Object[initialData.length];
        System.arraycopy(initialData, 0, this.data, 0, initialData.length);
    }

    // Search method
    public void search(T value) {
        for (int i = 0; i < data.length; i++) {
            if (data[i] != null && data[i].equals(value)) {
                System.out.println("Found: " + value + " at index " + i);
                return;
            }
        }
        System.out.println("Not found: " + value);
    }

    // Delete method
    @SuppressWarnings("unchecked")
    public void delete(T value) {
        int indexToDelete = -1;
        // Find the first occurrence of the value
        for (int i = 0; i < data.length; i++) {
            if (data[i] != null && data[i].equals(value)) {
                indexToDelete = i;
                break;
            }
        }

        if (indexToDelete == -1) {
            System.out.println("Not found, cannot delete: " + value);
            return;
        }

        // Create a new array with one less element
        T[] newData = (T[]) new Object[data.length - 1];
        int newIndex = 0;
        for (int i = 0; i < data.length; i++) {
            if (i != indexToDelete) {
                newData[newIndex] = data[i];
                newIndex++;
            }
        }
        data = newData;
        System.out.println("Deleted: " + value);
    }

    // Get current data
    public T[] getData() {
        return data;
    }

    // Main method for testing
    public static void main(String[] args) {
        // Example with Integers
        Integer[] intData = {4, 7, 6, 8, 34, 43, 65, 34};
        GenericArray<Integer> intArray = new GenericArray<>(intData);
        System.out.println("Initial Integer Array: " + Arrays.toString(intArray.getData()));
        intArray.search(6);
        intArray.search(99);
        intArray.delete(34);
        System.out.println("After deleting 34: " + Arrays.toString(intArray.getData()));

        // Example with Strings
        String[] strData = {"apple", "banana", "cherry", "date"};
        GenericArray<String> strArray = new GenericArray<>(strData);
        System.out.println("\nInitial String Array: " + Arrays.toString(strArray.getData()));
        strArray.search("banana");
        strArray.search("grape");
        strArray.delete("cherry");
        System.out.println("After deleting cherry: " + Arrays.toString(strArray.getData()));

        // Example with Doubles
        Double[] doubleData = {4.5, 7.2, 6.1, 8.9};
        GenericArray<Double> doubleArray = new GenericArray<>(doubleData);
        System.out.println("\nInitial Double Array: " + Arrays.toString(doubleArray.getData()));
        doubleArray.search(6.1);
        doubleArray.delete(7.2);
        System.out.println("After deleting 7.2: " + Arrays.toString(doubleArray.getData()));
    }
}

File: cpp/GenericArray.cpp
#include <iostream>
#include <string>
#include <algorithm>

template<typename T>
class GenericArray {
private:
    T* data; // Dynamic array to store elements
    size_t size; // Current size of the array

public:
    // Constructor to initialize the array
    GenericArray(const T* initialData, size_t initialSize) {
        size = initialSize;
        data = new T[size];
        std::copy(initialData, initialData + size, data);
    }

    // Destructor to free memory
    ~GenericArray() {
        delete[] data;
    }

    // Search method
    void search(const T& value) const {
        for (size_t i = 0; i < size; ++i) {
            if (data[i] == value) {
                std::cout << "Found: " << value << " at index " << i << std::endl;
                return;
            }
        }
        std::cout << "Not found: " << value << std::endl;
    }

    // Delete method
    void deleteValue(const T& value) {
        int indexToDelete = -1;
        // Find the first occurrence of the value
        for (size_t i = 0; i < size; ++i) {
            if (data[i] == value) {
                indexToDelete = i;
                break;
            }
        }

        if (indexToDelete == -1) {
            std::cout << "Not found, cannot delete: " << value << std::endl;
            return;
        }

        // Create a new array with one less element
        T* newData = new T[size - 1];
        size_t newIndex = 0;
        for (size_t i = 0; i < size; ++i) {
            if (i != static_cast<size_t>(indexToDelete)) {
                newData[newIndex] = data[i];
                ++newIndex;
            }
        }

        // Free old array and update to new array
        delete[] data;
        data = newData;
        --size;
        std::cout << "Deleted: " << value << std::endl;
    }

    // Print array contents
    void print() const {
        std::cout << "[";
        for (size_t i = 0; i < size; ++i) {
            std::cout << data[i];
            if (i < size - 1) {
                std::cout << ", ";
            }
        }
        std::cout << "]" << std::endl;
    }
};

int main() {
    // Example with integers
    int intData[] = {4, 7, 6, 8, 34, 43, 65, 34};
    size_t intSize = sizeof(intData) / sizeof(intData[0]);
    GenericArray<int> intArray(intData, intSize);
    std::cout << "Initial Integer Array: ";
    intArray.print();
    intArray.search(6);
    intArray.search(99);
    intArray.deleteValue(34);
    std::cout << "After deleting 34: ";
    intArray.print();

    // Example with doubles
    double doubleData[] = {4.5, 7.2, 6.1, 8.9};
    size_t doubleSize = sizeof(doubleData) / sizeof(doubleData[0]);
    GenericArray<double> doubleArray(doubleData, doubleSize);
    std::cout << "\nInitial Double Array: ";
    doubleArray.print();
    doubleArray.search(6.1);
    doubleArray.deleteValue(7.2);
    std::cout << "After deleting 7.2: ";
    doubleArray.print();

    // Example with strings
    std::string strData[] = {"apple", "banana", "cherry", "date"};
    size_t strSize = sizeof(strData) / sizeof(strData[0]);
    GenericArray<std::string> strArray(strData, strSize);
    std::cout << "\nInitial String Array: ";
    strArray.print();
    strArray.search("banana");
    strArray.search("grape");
    strArray.deleteValue("cherry");
    std::cout << "After deleting cherry: ";
    strArray.print();

    return 0;
}
